firebase-sessions/src/main/kotlin/com/google/firebase/sessions/SessionDataService.kt

```diff
@@ -37,65 +37,81 @@ import java.util.UUID
 
 /** Service for providing access to session data */
 internal class SessionDataService : Service() {
-  /** Target we publish for clients to send messages to IncomingHandler. */
-  private lateinit var messenger: Messenger
 
-  val boundClients = mutableListOf<Messenger>()
+  private val boundClients = mutableListOf<Messenger>()
+  private var curSessionId: String? = null
+  private var isAppInForeground: Boolean = false
+  private var lastMsgTimeMs: Long = 0
 
   /** Handler of incoming messages from clients. */
   internal inner class IncomingHandler(
     context: Context,
     private val appContext: Context = context.applicationContext,
   ) : Handler(Looper.getMainLooper()) { // TODO(rothbutter) probably want to use our own executor
 
-    private var curSessionId: String? = null
-
     override fun handleMessage(msg: Message) {
+      if (lastMsgTimeMs > msg.getWhen()) {
+        Log.i(TAG, "SERVICE: Received old message $msg. Ignoring")
+        return
+      }
       when (msg.what) {
-        FOREGROUNDED -> handleForegrounding()
-        BACKGROUNDED -> handleBackgrounding()
+        FOREGROUNDED -> handleForegrounding(msg)
+        BACKGROUNDED -> handleBackgrounding(msg)
         else -> super.handleMessage(msg)
       }
+      lastMsgTimeMs = msg.getWhen()
     }
+  }
 
-    fun handleForegrounding() {
-      Log.i(TAG, "SERVICE: Activity foregrounding - updating ${boundClients.size} clients")
-      broadcastSession(UUID.randomUUID().toString())
-    }
+  fun handleForegrounding(msg: Message) {
+    Log.i(TAG, "SERVICE: Activity foregrounding at ${msg.getWhen()}")
 
-    fun handleBackgrounding() {
-      Log.i(TAG, "SERVICE: Activity backgrounding")
+    if (curSessionId == null) {
+      Log.i(TAG, "SERVICE: Cold start detected.")
+      newSession()
+    } else if (!isAppInForeground && msg.getWhen() - lastMsgTimeMs > MAX_BACKGROUND_MS) {
```

Comment:
Why are we keeping track of the foreground state of the app? Is it to prevent multiple sessions from being sent? How about update the timestamp on a foreground instead?

Reply:
I think we need to know that the app is already in foreground for cases like split screens or multiple windows on tablets. So if an app is already foregrounded and we get another foreground event from a different activity, we don't want to trigger a new session if it has been 30m since the last foreground/background came in.